website/blog/2018-12-12-index.md

```diff
@@ -0,0 +1,290 @@
+
+To navigate to the `scala.Some.isEmpty` method we can't use the same approach as
+in Java because the class `scala.Some` is in `scala/Option.scala`, not
+`scala/Some.scala`.
+
+One possible solution is to read the `scala/Some.class` classfile that contains
+the filename `Option.scala` where `scala.Some` is defined.
+
+```sh
+$ javap -classpath $(coursier fetch org.scala-lang:scala-library:2.12.8) scala.Some
+Compiled from "Option.scala"
+...
+```
+
+However, source information in classfiles is not always reliable and it may be
+removed by tools that process jar files. Let's restrict the problem to only use
+source files.
+
+Instead of using JVM classfiles, we can walk all Scala source files and build an
+index that maps toplevel classes to the source file that defines that class. The
+index looks something like this:
+
+```scala
+val index = Map[Symbol, Path](
+  Symbol("scala.Some") -> Path("scala/Option.scala"),
+  // ...
+)
+```
+
+Our challenge is to build this index as fast as possible.
+
+## Initial solution
+
+The first implementation of the Metals indexer used the
+[Scalameta](https://scalameta.org/) parser. This parser does not desugar the
+original code making it useful for refactoring and code-formatting tools like
+[Scalafix](http://scalacenter.github.io/scalafix/)/[Scalafmt](http://scalameta.org/scalafmt).
+I'm also familiar with Scalameta parser API so it was fast to get a working
+implementation. However, is the parser fast enough to parse millions of lines of
+code on every server startup?
+
+According to JMH benchmarks, parsing a sizable corpus of Scala code, the
+Scalameta parser handles ~92k lines/second. The benchmarks use the "single-shot"
+mode of JMH, for which the documentation says:
+
+> "This mode is useful to estimate the "cold" performance when you don't want to
+> hide the warmup invocations."
+
+Cold performance is an OK estimate for our use-case since indexing happens
+during server startup.
+
+The Scala standard library is ~36k lines so at 92k lines/second this solution
+scales up to a codebase with up to 20-30 similarly-sized library dependencies.
+If we add more library dependencies, we exceed the 10 second constraint for
+indexing time. For a codebase with 5 million lines of code, users might have to
+wait one minute for indexing to complete. We should aim for better.
+
+## Optimized solution
+
+We can speed up indexing by writing a custom parser that extracts only the
+information we need from a source file. For example, the Scalameta parser
+extracts method implementations that are irrelevant for our indexer. Our indexer
+needs to know the toplevel classes and nothing more.
+
+The algorithm for this custom parser that only extracts toplevel classes goes
+roughly like this.
+
+- tokenize source file
+- on `package` keyword, record package name
+- on `class` and `trait` and `object` keywords, record toplevel class
+- on `(` and `[` and `{` delimiters, skip tokens until we find matching closing
+  delimiter
```

Comment:
What about braced package declarations:
```scala
package foo

package bar {
  class Foobar
}
```
Also: classes in `package object`s?
?

Reply:
Those are handled, the logic in this post is maybe a bit oversimplified. I'll update to make it clearer